### PR TITLE
FormatOps: improve binpacking parentConstructors

### PR DESCRIPTION
**Pull request by kitbellew**
Alternative to #3091.

### CODE REVIEW

File: scalafmt-tests/src/test/resources/binPack/ParentConstructors.stat [hunk 1]
```diff
@@ -114,3 +114,71 @@ def method[T <: ColType](): DTDataset[
...
+    val u: ru.type = ru
+    val mirror: u.Mirror = u.runtimeMirror(classOf[scala.meta.Tree].getClassLoader)
+  }
+  with scala.meta.internal.trees.Reflection
+  with scala.meta.internal.tokens.Reflection
+>>>
+object CoreReflection extends {
+      val u: ru.type = ru
+      val mirror: u.Mirror = u.runtimeMirror(classOf[scala.meta.Tree].getClassLoader)
+    } with scala.meta.internal.trees.Reflection with scala.meta.internal.tokens.Reflection
```
Comment (dos65) [inline scalafmt-tests/src/test/resources/binPack/ParentConstructors.stat:131]: it only fixes `extends` position.
early stats and `} with ..` shouldn't be indented

Comment (kitbellew) [reply scalafmt-tests/src/test/resources/binPack/ParentConstructors.stat:131]: why shouldn't they be indented? everything between the beginning of a template and the opening brace of its body should be indented.

in your case,
```
... {
      val u: ru.type = ru
      val mirror: u.Mirror = u.runtimeMirror(classOf[scala.meta.Tree].getClassLoader)
    }
```
is not the body of `CoreReflection` but of some other (parent) construct. definitely has to be indented.

Comment (kitbellew) [reply scalafmt-tests/src/test/resources/binPack/ParentConstructors.stat:131]: here's another example, existing behaviour:

https://github.com/scalameta/scalafmt/blob/master/scalafmt-tests/src/test/resources/binPack/ParentConstructors.stat#L49

your proposal is contrary to that.

File: scalafmt-tests/src/test/resources/scalajs/Class.stat [hunk 2]
```diff
@@ -54,8 +54,7 @@ private class Encoder extends CharsetEncoder(
     }
 >>>
 object a {
-  private class Encoder
-      extends CharsetEncoder(UTF_16_Common.this, 2.0f, 2.0f,
+  private class Encoder extends CharsetEncoder(UTF_16_Common.this, 2.0f, 2.0f,
```
Comment (tgodzik) [inline scalafmt-tests/src/test/resources/scalajs/Class.stat:57]: Is this change intended?

File: scalafmt-tests/src/test/resources/scala3/Enum.stat [hunk 3]
```diff
@@ -129,9 +129,8 @@ enum Foo(val x: Int) {
 }
 >>>
 enum Foo(val x: Int) {
-  case Bar
-      extends Foo(1) with X
-      with Y with Z
+  case Bar extends Foo(1)
```
Comment (tgodzik) [inline scalafmt-tests/src/test/resources/scala3/Enum.stat:132]: This seems wrong now. With `binPack.parentConstructors = Always` wasn't the previous state correct?

**Review by tgodzik — CHANGES_REQUESTED** [review]
The changes to early initializers look good, but I wonder if anything should have been also updated?

File: scalafmt-tests/src/test/resources/scalajs/Class.stat (hunk 2 above)
Comment (kitbellew) [reply scalafmt-tests/src/test/resources/scalajs/Class.stat:57]: yes, same explanation is above. `always` attempts to binpack as much as possible unless there was an explicit break.

File: scalafmt-tests/src/test/resources/binPack/ParentConstructors.stat (hunk 1 above)
Comment (dos65) [reply scalafmt-tests/src/test/resources/binPack/ParentConstructors.stat:131]: If we treat that `parentConstructor` doesn't affect early init body and work only for `extends`  and with chain it wouldn't be contrary.

Let me try to fix it it properly, I didn't noticed that there is such setting at the moment when I was preparing a fix.

Comment (kitbellew) [reply scalafmt-tests/src/test/resources/binPack/ParentConstructors.stat:131]: can you explain how init body is different? other than indentation, what else would you like to do?

**Comment by kitbellew** [comment]
just for visual effect, here's a version which avoids indenting early ctors: in my view, the formatting is not especially readable or pleasing, the dissonance with subsequent parents is prominent:

https://github.com/kitbellew/scalafmt/commit/ce42c5e343eae5d03f8ec7958cc6ed24550792b4

**Review by dos65 — APPROVED** [review]
:+1:
